### PR TITLE
feat: add Google OAuth2 Authentication for user login 

### PR DESCRIPTION
**Pull request by johnson-oragui**
​

This pull request implements Google OAuth2 authentication as a login option. This feature will allow users to log in using their Google accounts, making the authentication process seamless and secure. It enhances user convenience and security, reduces the need for password management, and offers a popular and trusted authentication method.
​

## Description

The changes this makes to the existing code-base is:
 - Created schemas for google oauth response
 - Created a configuration using authlib for google oauth
 - created google_oauth_services module to handle user creation for authenticated users via google oauth
 - create a route to handle the initial call to authenticate with google
 - created a callback-route that handles successful or un-successful response from google authentication server
 - updated main.py to include session management as required for the use of google oauth
 - updated the requirements.txt with google oauth dependencies/modules
 - wrote unittests(pytest) to mock the flow of google oauth (includes teardown)
 - wrote unittests(pytest) to confirm that data are been saved to the database after authentication (includes teardown)

​

## Related Issue
[Google OAuth2 Authentication for User Login](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/33)

​

## Motivation and Context

Google oauth is currently a consistent way users all over different web applications use to access a resource(API or UI) seamlessly. Not having it as a feature to this would make the boilerplate look like it time travelled not into the future, but back into the past. It is now a necessity to have a feature in applications for Oauth to be present. With this feature, a user does not need to go through the hassle of authenticating or accessing a resource by creating new passwords and memorizing it, the user would easily login with an already existing account on google.

​

## How Has This Been Tested?

I Tested the google oauth login route using my browser, I started the application locally, opened the link on my browser, the link redirected me to the login screen of google, I entered my details, it redirected me to the callback route, and my details were saved to the database, and in return I received an access token and a refresh token with user data. The changes  made to the boilerplate did not affect any other part of the code, as it solely depends on just using the create_access_token function already created.

## Screenshots (if appropriate - Postman, etc):

​

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
      ​

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Tasks Completed
- [x] Google OAuth2 Login Endpoint: Redirects the user to Google's OAuth2 login page and handles the OAuth2 flow.
- [x] Callback Endpoint: Handles the OAuth2 callback, exchanges authorization code for tokens, and fetches user information.
- [x] Token Management: Issues JWT access and refresh tokens and stores them in relation to the user.
- [x] User Information Update: Fetches and updates user information from Google on re-authentication.
- [x] Error Handling: Properly manages errors related to authentication, token exchange, and user information fetching.

## Expected Outcome
 - The user will be able to easily authenticate with their Google account, receive a token and a refresh token, and use the service with the issued tokens. The system will handle user information updates and error scenarios effectively.

Please review the changes and let me know if there are any adjustments or additional features required. Thank you!

### CODE REVIEW

**Comment by CodewithSegNet** [comment]
this is good.

**Review by CodewithSegNet — APPROVED** [review]
looks good. however you'd have to wait for a mentor review

**Comment by johnson-oragui** [comment]
> looks good. however you'd have to wait for a mentor review

Thank you. I will wait.

**Comment by CodewithSegNet** [comment]
wrong PR name though. follow the proper PR naming convention. eg, feat:

**Comment by johnson-oragui** [comment]
> wrong PR name though. follow the proper PR naming convention. eg, feat:

Thank you. I have changed it from "FEAT" to "feat"

**Comment by CodewithSegNet** [comment]
remove the [] and add a : after feat
eg, feat:

**Comment by johnson-oragui** [comment]
> remove the [] and add a : after feat eg, feat:

again. Thank you. Change done.

File: api/v1/schemas/google_oauth.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+class UserData(BaseModel):
+    """
+    Schema Response representing the validated google login
+    """
+    id: str
+    first_name: str
+    last_name: str
+    username: str
+    email: str
```
Comment (zxenonx) [inline api/v1/schemas/google_oauth.py:12]:
```suggestion
    email: EmailStr
```
needs to be imported from Pydantic

**Review by zxenonx — COMMENTED** [review]
Thanks @johnson-oragui 
I left a comment

File: api/v1/schemas/google_oauth.py (hunk 1 above)
Comment (zxenonx) [reply api/v1/schemas/google_oauth.py:12]: The screenshots section is empty... @johnson-oragui

**Comment by buka4rill** [comment]
Wrong Branch name. Closed

File: api/v1/schemas/google_oauth.py (hunk 1 above)
Comment (johnson-oragui) [reply api/v1/schemas/google_oauth.py:12]: > The screenshots section is empty... @johnson-oragui

It's difficult testing it with postman, postman uses a different Oauth flow from the Authlib module. But I will add the screenshots for testing with the browser, and keep trying with postman.

Thank you.

**Pull request closed without merging**